File: src/main/scala/chapter6/scala/WindowFunctions.scala
package chapter6.scala

import org.apache.flink.api.common.functions.{AggregateFunction, MapFunction, ReduceFunction}
import org.apache.flink.streaming.api.datastream.DataStream
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.windows.TimeWindow
import org.apache.flink.util.Collector
import util.scala.{SensorReading, SensorSource, WatermarkStrategies}

import java.lang

object WindowFunctions {

  def threshold = 25.0

  def main(args: Array[String]): Unit = {

    // set up the streaming execution environment
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    // checkpoint every 10 seconds
    env.getCheckpointConfig.setCheckpointInterval(10 * 1000)

    // ingest sensor stream
    val sensorData: DataStream[SensorReading] = env
      .addSource(new SensorSource)
      .assignTimestampsAndWatermarks(WatermarkStrategies.sensorReadingStrategy)

    val minTempPerWindow: DataStream[(String, Double)] = sensorData
      .map(new MapFunction[SensorReading, (String, Double)] {
        override def map(t: SensorReading): (String, Double) = (t.id, t.temperature)
      })
      .keyBy((in: (String, Double)) => in._1)
      .window(TumblingEventTimeWindows.of(Time.seconds(15)))
      .reduce((r1, r2) => (r1._1, r1._2 min r2._2))

    val minTempPerWindow2: DataStream[(String, Double)] = sensorData
      .map(new MapFunction[SensorReading, (String, Double)] {
        override def map(t: SensorReading): (String, Double) = (t.id, t.temperature)
      })
      .keyBy((in: (String, Double)) => in._1)
      .window(TumblingEventTimeWindows.of(Time.seconds(15)))
      .reduce(new MinTempFunction)

    val avgTempPerWindow: DataStream[(String, Double)] = sensorData
      .map(new MapFunction[SensorReading, (String, Double)] {
        override def map(t: SensorReading): (String, Double) = (t.id, t.temperature)
      })
      .keyBy((in: (String, Double)) => in._1)
      .window(TumblingEventTimeWindows.of(Time.seconds(15)))
      .aggregate(new AvgTempFunction)

    // output the lowest and highest temperature reading every 5 seconds
    val minMaxTempPerWindow: DataStream[MinMaxTemp] = sensorData
      .keyBy((in: SensorReading) => in.id)
      .window(TumblingEventTimeWindows.of(Time.seconds(5)))
      .process(new HighAndLowTempProcessFunction)

    val minMaxTempPerWindow2: DataStream[MinMaxTemp] = sensorData
      .map(new MapFunction[SensorReading, (String, Double, Double)] {
        override def map(t: SensorReading): (String, Double, Double) = (t.id, t.temperature, t.temperature)
      })
      .keyBy((in: (String, Double, Double)) => in._1)
      .window(TumblingEventTimeWindows.of(Time.seconds(5)))
      .reduce(
        // incrementally compute min and max temperature
        (r1: (String, Double, Double), r2: (String, Double, Double)) => {
          (r1._1, r1._2 min r2._2, r1._3 max r2._3)
        },
        // finalize result in ProcessWindowFunction
        new AssignWindowEndProcessFunction()
      )

    // print result stream
    minMaxTempPerWindow2.print()

    env.execute()
  }
}

class MinTempFunction extends ReduceFunction[(String, Double)] {
  override def reduce(r1: (String, Double), r2: (String, Double)): (String, Double) = (r1._1, r1._2 min r2._2)
}

class AvgTempFunction extends AggregateFunction[(String, Double), (String, Double, Int), (String, Double)] {

  override def createAccumulator(): (String, Double, Int) = ("", 0.0, 0)

  override def add(in: (String, Double), acc: (String, Double, Int)): (String, Double, Int) = (in._1, in._2 + acc._2, 1 + acc._3)

  override def getResult(acc: (String, Double, Int)): (String, Double) = (acc._1, acc._2 / acc._3)

  override def merge(acc1: (String, Double, Int), acc2: (String, Double, Int)): (String, Double, Int) = (acc1._1, acc1._2 + acc2._2, acc1._3 + acc2._3)
}

case class MinMaxTemp(id: String, min: Double, max:Double, endTs: Long)

/**
 * A ProcessWindowFunction that computes the lowest and highest temperature
 * reading per window and emits a them together with the
 * end timestamp of the window.
 */
class HighAndLowTempProcessFunction extends ProcessWindowFunction[SensorReading, MinMaxTemp, String, TimeWindow] {

  override def process(key: String,
                       ctx: ProcessWindowFunction[SensorReading, MinMaxTemp, String, TimeWindow]#Context,
                       vals: lang.Iterable[SensorReading],
                       collector: Collector[MinMaxTemp]): Unit = {

    val temps = collection.mutable.ArrayBuffer[Double]()
    val it = vals.iterator()
    while (it.hasNext) temps += it.next().temperature

    val windowEnd = ctx.window.getEnd

    collector.collect(MinMaxTemp(key, temps.min, temps.max, windowEnd))

  }
}

class AssignWindowEndProcessFunction extends ProcessWindowFunction[(String, Double, Double), MinMaxTemp, String, TimeWindow] {

  override def process(key: String,
                       ctx: ProcessWindowFunction[(String, Double, Double), MinMaxTemp, String, TimeWindow]#Context,
                       iterable: lang.Iterable[(String, Double, Double)],
                       collector: Collector[MinMaxTemp]): Unit = {

    val minMax = iterable.iterator().next()
    val windowEnd = ctx.window.getEnd
    collector.collect(MinMaxTemp(key, minMax._2, minMax._3, windowEnd))
  }
}

File: src/main/java/chapter6/WindowFunctions.java
package chapter6;

import org.apache.flink.api.common.functions.AggregateFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.common.functions.ReduceFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.util.Collector;
import util.SensorReading;
import util.SensorSource;
import util.WatermarkStrategies;

import java.util.ArrayDeque;
import java.util.Comparator;
import java.util.Iterator;

public class WindowFunctions {

    double threshold = 25.0;

    public static void main(String[] args) throws Exception {

    // set up the streaming execution environment
    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

    // checkpoint every 10 seconds
    env.getCheckpointConfig().setCheckpointInterval(10 * 1000);

    // ingest sensor stream
    DataStream<SensorReading> sensorData = env
      .addSource(new SensorSource())
      .assignTimestampsAndWatermarks(new WatermarkStrategies().sensorReadingStrategy);

    DataStream<Tuple2<String, Double>> minTempPerWindow = sensorData
        .map(new MapFunction<SensorReading, Tuple2<String, Double>>() {
            @Override
            public Tuple2<String, Double> map(SensorReading sensorReading) throws Exception {
                return Tuple2.of(sensorReading.getId(), sensorReading.getTemperature());
            }
        })
        .keyBy(in -> in.f0)
        .window(TumblingEventTimeWindows.of(Time.seconds(15)))
        .reduce((r1, r2) -> Tuple2.of(r1.f0, Math.min(r1.f1, r2.f1)));

    DataStream<Tuple2<String, Double>> minTempPerWindow2 = sensorData
        .map(new MapFunction<SensorReading, Tuple2<String, Double>>() {
            @Override
            public Tuple2<String, Double> map(SensorReading sensorReading) throws Exception {
                return Tuple2.of(sensorReading.getId(), sensorReading.getTemperature());
            }
        })
        .keyBy(in -> in.f0)
        .window(TumblingEventTimeWindows.of(Time.seconds(15)))
        .reduce(new MinTempFunction());

    DataStream<Tuple2<String, Double>> avgTempPerWindow = sensorData
        .map(new MapFunction<SensorReading, Tuple2<String, Double>>() {
            @Override
            public Tuple2<String, Double> map(SensorReading sensorReading) throws Exception {
                return Tuple2.of(sensorReading.getId(), sensorReading.getTemperature());
            }
        })
        .keyBy(in -> in.f0)
        .window(TumblingEventTimeWindows.of(Time.seconds(15)))
        .aggregate(new AvgTempFunction());

    // output the lowest and highest temperature reading every 5 seconds
    DataStream<MinMaxTemp> minMaxTempPerWindow = sensorData
        .keyBy(SensorReading::getId)
        .window(TumblingEventTimeWindows.of(Time.seconds(5)))
        .process(new HighAndLowTempProcessFunction());

    DataStream<MinMaxTemp> minMaxTempPerWindow2 = sensorData
        .map(new MapFunction<SensorReading, Tuple3<String, Double, Double>>() {
            @Override
            public Tuple3<String, Double, Double> map(SensorReading sensorReading) throws Exception {
                return Tuple3.of(sensorReading.getId(), sensorReading.getTemperature(), sensorReading.getTemperature());
            }
        })
        .keyBy(in -> in.f0)
        .window(TumblingEventTimeWindows.of(Time.seconds(5)))
        .reduce(
            (r1, r2) -> Tuple3.of(r1.f0, Math.min(r1.f1, r2.f1), Math.max(r1.f2, r2.f2)),
            new AssignWindowEndProcessFunction()
        );

    // print result stream
    minMaxTempPerWindow2.print();

    env.execute();
  }
}

class MinTempFunction implements ReduceFunction<Tuple2<String, Double>> {
    @Override
    public Tuple2<String, Double> reduce (Tuple2<String, Double> r1, Tuple2<String, Double> r2) {
        return Tuple2.of(r1.f0, Math.min(r1.f1, r2.f1));
    }

}

class AvgTempFunction implements AggregateFunction<Tuple2<String, Double>, Tuple3<String, Double, Long>, Tuple2<String, Double>> {

    @Override
    public Tuple3<String, Double, Long> createAccumulator() {
        return Tuple3.of("", 0.0, 0L);
    }

    @Override
    public Tuple3<String, Double, Long> add(Tuple2<String, Double> stringDoubleTuple2, Tuple3<String, Double, Long> stringDoubleLongTuple3) {
        return Tuple3.of(stringDoubleLongTuple3.f0, stringDoubleLongTuple3.f1 + stringDoubleTuple2.f1, stringDoubleLongTuple3.f2 + 1);
    }

    @Override
    public Tuple2<String, Double> getResult(Tuple3<String, Double, Long> stringDoubleLongTuple3) {
        return Tuple2.of(stringDoubleLongTuple3.f0, stringDoubleLongTuple3.f1 / stringDoubleLongTuple3.f2);
    }

    @Override
    public Tuple3<String, Double, Long> merge(Tuple3<String, Double, Long> acc1, Tuple3<String, Double, Long> acc2) {
        return Tuple3.of(acc1.f0, acc1.f1 + acc2.f1, acc1.f2 + acc2.f2);
    }
}

class MinMaxTemp {
    public String id;
    public Double min;
    public Double max;
    public Long endTs;

    @Override
    public String toString() {
        return "MinMaxTemp{" +
            "id='" + id + '\'' +
            ", min=" + min +
            ", max=" + max +
            ", endTs=" + endTs +
            '}';
    }

    public MinMaxTemp(String id, Double min, Double max, Long endTs) {
        this.id = id;
        this.min = min;
        this.max = max;
        this.endTs = endTs;
    }
}

/**
 * A ProcessWindowFunction that computes the lowest and highest temperature
 * reading per window and emits a them together with the
 * end timestamp of the window.
 */
class HighAndLowTempProcessFunction extends ProcessWindowFunction<SensorReading, MinMaxTemp, String, TimeWindow> {

    @Override
    public void process(String key,
                        ProcessWindowFunction<SensorReading, MinMaxTemp, String, TimeWindow>.Context context,
                        Iterable<SensorReading> iterable,
                        Collector<MinMaxTemp> collector) throws Exception {

        Iterator<SensorReading> it = iterable.iterator();
        ArrayDeque<Double> deque = new ArrayDeque<>();

        while (it.hasNext()) deque.add(it.next().getTemperature());

        Long windowEnd = context.window().getEnd();

        collector.collect(new MinMaxTemp(key, deque.stream().min(Comparator.naturalOrder()).get(), deque.stream().max(Comparator.naturalOrder()).get(), windowEnd));

    }
}

class AssignWindowEndProcessFunction extends ProcessWindowFunction<Tuple3<String, Double, Double>, MinMaxTemp, String, TimeWindow> {

    @Override
    public void process(String key,
                        ProcessWindowFunction<Tuple3<String, Double, Double>, MinMaxTemp, String, TimeWindow>.Context context,
                        Iterable<Tuple3<String, Double, Double>> iterable,
                        Collector<MinMaxTemp> collector) throws Exception {

      Tuple3<String, Double, Double> minMax = iterable.iterator().next();
      Long windowEnd = context.window().getEnd();
      collector.collect(new MinMaxTemp(key, minMax.f1, minMax.f2, windowEnd));
    }
}